### PR TITLE
String-typed runtime in-filter referene freed key column of hash table

### PR DESCRIPTION
**Pull request by satanson**
## What type of PR is this：
- [x] bug
- [ ] feature
- [ ] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #
https://github.com/StarRocks/starrocks/issues/4183
https://github.com/StarRocks/starrocks/issues/4066

## Problem Summary(Required) ：

String typed runtim in-filter can reference freed key column of hash table, hashset of string-typed in-filter keeps Slice. 
In serveral cases, one pair of probe side PipelineDriver and build side PipelineDriver will finalize before the last HashJoinBuildOperator merge in-filters when degree-of-parallelism > 1.

case1: enable_global_runtime_filter = false and enable_pipeline_engine=true,  then Global RuntimeFilters are not planned, so local_rf_waiting_set based on Global RuntimeFilters are not generated, in pipeline engine, Operator use local_rf_waiting_set to determine which HashJoinBuildOperators it should wait for until the HashJoinBuildOperators have completed runtime-filter generation and merging. if the local rf_wating_set is empty, then probe side will be executed in advance before runtime-filters are merged in the last HashJoinBuildOperator::set_finishing invocation.

case2: in shuffle join, even if enable_global_runtime_filter = true, Global RuntimeFilters are constructed but they have no remote targets, this type GRFs are not used to generate local_rf_waiting_set, but in BE, in case of build_expr_contexts is not empty, the in-filters are also constructed. so ExchangeSourceOperators will run in advance because of empty set of local_rf_waiting_set.

### CODE REVIEW

**Review by murphyatwork — DISMISSED** [review]

File: be/src/exec/vectorized/hash_joiner.cpp [hunk 1]
```diff
@@ -257,7 +257,7 @@ Status HashJoiner::create_runtime_filters(RuntimeState* state) {
             _is_push_down = false;
         }
 
-        if (_is_push_down || _build_conjunct_ctxs_is_empty) {
+        if (_is_push_down || !_build_conjunct_ctxs_is_empty) {
```
Comment (trueeyu) [inline be/src/exec/vectorized/hash_joiner.cpp:260]: fix another bug of not push down ?

Comment (satanson) [reply be/src/exec/vectorized/hash_joiner.cpp:260]: yes

**Review by trueeyu — DISMISSED** [review]

**Review by murphyatwork — DISMISSED** [review]

**Comment by satanson** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 25 / 39 (64.10%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/HashJoinNode.java|0|2|00.00%|
|:large_blue_circle:|com/starrocks/planner/PlanNode.java|9|18|50.00%|
|:large_blue_circle:|com/starrocks/planner/PlanFragment.java|6|9|66.67%|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|10|10|100.00%|

**Comment by satanson** [comment]
@Mergifyio rebase

**Comment by satanson** [comment]
@Mergifyio rebase

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 26 / 39 (66.67%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/HashJoinNode.java|0|2|00.00%|
|:large_blue_circle:|com/starrocks/planner/PlanNode.java|10|18|55.56%|
|:large_blue_circle:|com/starrocks/planner/PlanFragment.java|6|9|66.67%|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|10|10|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 31 / 44 (70.45%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/HashJoinNode.java|0|2|00.00%|
|:large_blue_circle:|com/starrocks/planner/PlanNode.java|11|19|57.89%|
|:large_blue_circle:|com/starrocks/planner/PlanFragment.java|10|13|76.92%|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|10|10|100.00%|

**Comment by satanson** [comment]
@Mergifyio rebase

**Comment by satanson** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 31 / 44 (70.45%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/HashJoinNode.java|0|2|00.00%|
|:large_blue_circle:|com/starrocks/planner/PlanNode.java|11|19|57.89%|
|:large_blue_circle:|com/starrocks/planner/PlanFragment.java|10|13|76.92%|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|10|10|100.00%|

File: fe/fe-core/src/main/java/com/starrocks/planner/PlanFragment.java [hunk 2]
```diff
@@ -233,6 +236,24 @@ public void setDopEstimated() {
...
+            hashJoinNodeIds.add(root.getId().asInt());
+        }
+        if (clearGlobalRuntimeFilter) {
+            root.clearProbeRuntimeFilters();
+            if (root instanceof HashJoinNode) {
+                ((HashJoinNode) root).clearBuildRuntimeFilters();
+            }
+        }
+        for (PlanNode child : root.getChildren()) {
+            if (child.getFragment() == this) {
```
Comment (ZiheLiu) [inline fe/fe-core/src/main/java/com/starrocks/planner/PlanFragment.java:251]: Maybe `if (child.getFragment() != this)`?

File: fe/fe-core/src/main/java/com/starrocks/planner/PlanFragment.java [hunk 3]
```diff
@@ -233,6 +236,24 @@ public void setDopEstimated() {
...
+                computeLocalRfWaitingSet(child, clearGlobalRuntimeFilter);
+            }
```
Comment (ZiheLiu) [inline fe/fe-core/src/main/java/com/starrocks/planner/PlanFragment.java:253]: Maybe if `(child.getFragment() != this)`?

Comment (satanson) [reply fe/fe-core/src/main/java/com/starrocks/planner/PlanFragment.java:253]: child.getFragment() == this means only tackle child planNode in the same fragment instance.  here is collecting the local runtime filter.

**Review by murphyatwork — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Review by trueeyu — APPROVED** [review]

**Pull request merged**